pkg/webhook/admission/builder/builder.go

```diff
@@ -0,0 +1,202 @@
+
+	// This field maps to the FailurePolicy in the admissionregistrationv1beta1.Webhook
+	failurePolicy *admissionregistrationv1beta1.FailurePolicyType
+
+	// This field maps to the NamespaceSelector in the admissionregistrationv1beta1.Webhook
+	namespaceSelector *metav1.LabelSelector
+
+	// manager is the manager for the webhook.
+	manager manager.Manager
+}
+
+// NewWebhookBuilder creates an empty WebhookBuilder.
+func NewWebhookBuilder() *WebhookBuilder {
+	return &WebhookBuilder{}
+}
+
+// Name sets the Name of the webhook.
+// This is optional
+func (b *WebhookBuilder) Name(name string) *WebhookBuilder {
+	b.name = name
+	return b
+}
+
+// Type sets the type of the admission webhook
+// This is required
+func (b *WebhookBuilder) Type(t types.WebhookType) *WebhookBuilder {
+	b.t = &t
+	return b
+}
+
+// Path sets the Path for the webhook.
+// This is optional
+func (b *WebhookBuilder) Path(path string) *WebhookBuilder {
+	b.path = path
+	return b
+}
+
+// Operations sets the operations that this webhook cares.
+// It will be overridden by Rules if Rules are not empty.
+// This is optional
+func (b *WebhookBuilder) Operations(ops ...admissionregistrationv1beta1.OperationType) *WebhookBuilder {
+	b.operations = ops
+	return b
+}
+
+// ForType sets the type of resources that the webhook will operate.
+// This cannot be use with Rules.
+func (b *WebhookBuilder) ForType(obj runtime.Object) *WebhookBuilder {
+	b.apiType = obj
+	return b
+}
+
+// Rules sets the RuleWithOperations for the webhook.
+// It overrides ForType and Operations.
+// This is optional and for advanced user
+func (b *WebhookBuilder) Rules(rules ...admissionregistrationv1beta1.RuleWithOperations) *WebhookBuilder {
+	b.rules = rules
+	return b
+}
+
+// FailurePolicy sets the FailurePolicy of the webhook.
+// If not set, it will be defaulted by the server.
+// This is optional
+func (b *WebhookBuilder) FailurePolicy(policy admissionregistrationv1beta1.FailurePolicyType) *WebhookBuilder {
+	b.failurePolicy = &policy
+	return b
+}
+
+// NamespaceSelector sets the NamespaceSelector for the webhook.
+// This is optional
+func (b *WebhookBuilder) NamespaceSelector(namespaceSelector *metav1.LabelSelector) *WebhookBuilder {
+	b.namespaceSelector = namespaceSelector
+	return b
+}
+
+// WithManager set the manager for the webhook for provisioning client etc.
+func (b *WebhookBuilder) WithManager(mgr manager.Manager) *WebhookBuilder {
+	b.manager = mgr
+	return b
+}
+
+func (b *WebhookBuilder) validate() error {
+	if b.t == nil {
+		return errors.New("webhook type cannot be nil")
+	}
+	if b.rules == nil && b.apiType == nil {
+		return fmt.Errorf("ForType should be set")
+	}
+	if b.rules != nil && b.apiType != nil {
+		return fmt.Errorf("at most one of ForType and Rules can be set")
+	}
+	return nil
+}
+
+// Build creates the Webhook based on the options provided.
+func (b *WebhookBuilder) Build(handlers ...admission.Handler) (*admission.Webhook, error) {
```

Comment:
Added a `setDefaults` method.
I didn't add `admission.NewWebhook(webhookoptions)` since `webhookoptions` will be almost the same as `Webhook` struct. I'd like to let the users to use `Builder` or directly construct the `Webhook` struct. Then we will do the possible defaulting for `Webhook`.